src/dftbp/common/schedule.F90

```diff
@@ -177,4 +227,144 @@ subroutine getChunkRanges(groupSize, myRank, globalFirst, globalLast, localFirst
   end subroutine getChunkRanges
 
 
+  !> Calculate the chunk ranges for a given MPI-communicator considerung different workload
+  subroutine getChunkIterWithWorkload(groupSize, myRank, globalFirst, globalLast, workload, chunkIter)
+
+    !> Size of the group over which the chunks should be distributed
+    integer, intent(in) :: groupSize
+
+    !> Rank of the current process
+    integer, intent(in) :: myRank
+
+    !> First element of the range
+    integer, intent(in) :: globalFirst
+
+    !> Last element of the range
+    integer, intent(in) :: globalLast
+
+    !> Workload for each item
+    integer, intent(in) :: workload(:)
+
+    !> The chunk iterator
+    type(TChunkIterator), intent(out) :: chunkIter
+
+    integer :: numIndices, rank, i
+    integer, allocatable :: rankWorkload(:), indices(:)
+
+    allocate(rankWorkload(groupSize))
+    allocate(indices(globalLast - globalFirst + 1))
+
+    rankWorkload(:) = 0
+    indices(:) = 0
+    numIndices = 0
+
+    do i = globalFirst, globalLast
+      rank = minloc(rankWorkload, dim=1)
+      rankWorkload(rank) = rankWorkload(rank) + max(1, workload(i))
+      if (rank == myRank + 1) then
+        numIndices = numIndices + 1
+        indices(numIndices) = i
+      end if
+    end do
+
+    call TChunkIterator_init(chunkIter, indices(1:numIndices))
+
+  end subroutine getChunkIterWithWorkload
+
+
+  !> Initializes a new chunk iterator
+  subroutine TChunkIterator_init(this, indices)
+
+    !> Instance
+    class(TChunkIterator), intent(out) :: this
+
+    !> Index list
+    integer, intent(in) :: indices(:)
+
+    @:ASSERT(size(indices) > 0)
+    @:ASSERT(all(indices > 0))
+
+    allocate(this%indices(size(indices)))
```

Comment:
This is line is superfluous. The next line uses the automatic reallocation of the LHS, so that `this%indices` will be deallocated and reallocated in the next line again. (Alternatively, you might use `this%indices(:) = indices` in the next line, but why using two lines for something, which can be written also in one....)